vizro-core/src/vizro/models/_components/ag_grid.py

```diff
@@ -111,7 +111,7 @@ def build(self):
                 # here) must have the same setting as the object that is built by the on-page-load mechanism using
                 # with the user settings and rendered finally. Otherwise the grid is not rendered correctly.
                 # Hence be careful when editing the line below.
-                html.Div(self.__call__(data_frame=pd.DataFrame()), id=self.id, className="table-container"),
+                html.Div(self.__call__(), id=self.id, className="table-container"),
```

Comment:
Maybe something like this could work:

```
 def build(self):
        clientside_callback(...)
        grid = self.__call__()
        grid["rowData"] = {}  # or `grid["rowData"] = pd.DataFrame()`
        return dcc.Loading([..., html.Div(grid, ...), ...], ...)
```

This should send the original data_frame into the `figure` callable, but also return an empty DataFrame with all calculated configuration on the `build load`. Then the UI chart goes into the Loading state , and finally the `update_figures` returns filtered data + parametrised configuration.

Reply:
It is an interesting idea that I will try out

Reply:
I tried this in https://github.com/mckinsey/vizro/pull/439/commits/4c3924653c2db277d9200dbd010005d043a1a6f6, and it _almost works_.

There seem to be no issues with pagination for example, but unfortunately it seems to break the persistence settings.

Presumably that is because it "forgets" the filters when an intermediate AGGrid has no data. You can check out the complete changes at the commit above. Try setting a filter for the first AGGrid and change page. Upon returning the filter is gone, although persistence is set to true.

Reply:
Ah ok, thanks for trying it out! I'm fine with your current solution then. Hopefully we can improve this some time in the future.